File: 0790-domino-and-tromino-tiling/0790-domino-and-tromino-tiling.cpp
class Solution {
public:
    int numTilings(int n) {
        int mod=1e9+7;
        long long dp1 = 1, dp2 = 2, dp3 = 5, dp4;
        if(n<=3){
            switch(n){
                case 1: return dp1;
                case 2: return dp2;
                case 3: return dp3; 
            }
        }
        for (int i=4; i<=n; i++) {
            dp4 = (2* dp3 + dp1) % mod;
            dp1 = dp2;
            dp2 = dp3;
            dp3 = dp4;
        }
        return dp4;
    }
};

File: 0790-domino-and-tromino-tiling/0790-domino-and-tromino-tiling.java
class Solution {
    public int numTilings(int n) {
        int MOD = 1_000_000_007;
        long[] dp = new long[4];
        dp[0] = 1;
        dp[1] = 2;
        dp[2] = 5;
        if (n < 4)
            return (int) dp[n - 1];
        for (int i = 4; i <= n; i++) {
            dp[3] = (2 * dp[2] + dp[0]) % MOD;
            for (int j = 0; j < 3; j++)
                dp[j] = dp[j + 1];
        }
        return (int) dp[3];
    }
}